<a href="https://colab.research.google.com/github/IsabelenaVB/Business-Intelligence/blob/main/4_All.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**NAIVE BAYES**

In [ ]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

# Cargar el archivo Excel
nxl = "/content/drive/MyDrive/SEMESTRE-6/4. BD2_NBA_Data_M_Int.xlsx"
XDB = pd.read_excel(nxl, sheet_name=0)

# Seleccionar las variables
XDB = XDB[["salary", "games_played_0y1", "minutes_per_game"]].dropna()

# Separar las variables de entrada y salida
XD = XDB[["salary", "minutes_per_game"]]
yd = XDB["games_played_0y1"]

# Modelo Naive Bayes
mnb = GaussianNB()
mnb.fit(XD, yd)

# Obtener información del modelo
u = mnb.theta_
print("La media de las variables es:", np.round(u, 2))
var = mnb.var_
print("La varianza de las variables es:", np.round(var, 2))
sigma = np.sqrt(var)
print("La desviación estándar es:", np.round(sigma, 2))

# Determinar los límites
LS = u + sigma
LI = u - sigma
print("El límite superior es:", np.round(LS, 2))
print("El límite inferior es:", np.round(LI, 2))


# Crear la matriz de confusión
y_pred = mnb.predict(XD)
cm = confusion_matrix(yd, y_pred)
print("La matriz de confusión es:\n", cm)

# Métricas de desempeño
exactitud = (cm[0,0] + cm[1,1]) / cm.sum()
precision = cm[1,1] / (cm[1,1] + cm[0,1]) if (cm[1,1] + cm[0,1]) != 0 else 0
sensibilidad = cm[1,1] / (cm[1,1] + cm[1,0]) if (cm[1,1] + cm[1,0]) != 0 else 0
especificidad = cm[0,0] / (cm[0,0] + cm[0,1]) if (cm[0,0] + cm[0,1]) != 0 else 0

print("La exactitud es:", exactitud)
print("La precisión es:", precision)
print("La sensibilidad es:", sensibilidad)
print("La especificidad es:", especificidad)


La media de las variables es: [[2.91582743e+06 1.82900000e+01]
 [7.42986398e+06 2.67000000e+01]]
La varianza de las variables es: [[1.15798849e+13 2.74045400e+04]
 [3.68265886e+13 2.73810100e+04]]
La desviación estándar es: [[3.40292300e+06 1.65540000e+02]
 [6.06849146e+06 1.65470000e+02]]
El límite superior es: [[6.31875042e+06 1.83830000e+02]
 [1.34983554e+07 1.92170000e+02]]
El límite inferior es: [[-4.87095570e+05 -1.47260000e+02]
 [ 1.36137252e+06 -1.38770000e+02]]
La matriz de confusión es:
 [[234  21]
 [111  79]]
La exactitud es: 0.7033707865168539
La precisión es: 0.79
La sensibilidad es: 0.41578947368421054
La especificidad es: 0.9176470588235294


#**ARBOL DE DECISIONES**

In [ ]:
from ast import Import
#Librerias cientificas
import numpy as np
import pandas as pd

#Librerias Arbol
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

#Conctar al Drive
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
#Este paso es cargar la ruta de mi documento SUPER IMPORTANTE
nxl="/content/drive/MyDrive/SEMESTRE-6/4. BD2_NBA_Data_M_Int.xlsx"

#Antes de usar el dropna se seleccionan las variables para no perder info
XDB = pd.read_excel(nxl, sheet_name=0)  # Cargar la base de datos
XD = XDB[["salary", "age", "height", "weight", "games_started", "minutes_per_game", "assists", "blocks", "steals"]]
yd= XDB[["games_played_0y1"]]

XDB=XDB.dropna()
display(XDB)

#Modelo de Arbol
mar=DecisionTreeClassifier(criterion="gini", max_depth=4) #Ponerle maximo 4 niveles
mar.fit(XD,yd) #Ajustar el modelo de arbol entre las variables de entrada y la salida

#Se procede a hacer la grafica
from pydotplus import graph_from_dot_data #Esta opcion me crea la grafica
from sklearn.tree import export_graphviz #Esta opcion me exporta la grafica a PNG, JPEG, TIFF

ve=["salary", "age", "height", "weight", "games_started", "minutes_per_game", "assists", "blocks", "steals"] #Nombres para el arbol
dot_data=export_graphviz(mar, feature_names=ve) #Crear el grafico
graph=graph_from_dot_data(dot_data)
graph.write_png("ArbolDecision.png")

# Cargar la base de datos a pronosticar
XDB0 = pd.read_excel(nxl, sheet_name=0)  # Cargar la base de datos nuevamente
XDB0 = XDB0.dropna()
XD0 = XDB0[["salary", "age", "height", "weight", "games_started", "minutes_per_game", "assists", "blocks", "steals"]]
yd0 = XDB0["games_played_0y1"]
# Determinar las filas sin games_played
filas = np.where(pd.isna(yd0))[0]
if len(filas) > 0:
    XD0 = XD0.iloc[filas].to_numpy()

    # Calcular predicciones
    y_pred = mar.predict(XD0)

ydp=mar.predict(XD) #Pronostico sobre la base de datos original
cm=confusion_matrix(yd, ydp) #Matriz de confusion
print("La matriz de confusion es:\n",cm)

VN=cm[0,0] #Verdaderos Negativos
VP=cm[1,1] #Verdaderos Positivos
FN=cm[1,0] #Falsos Negativos
FP=cm[0,1] #Falsos Positivos

exactitud=(VP+VN)/(VP+VN+FP+FN)
print("La exactitud del modelo es: ",exactitud)
sensibilidad=VP/(VP+FN)
print("La sensibilidad del modelo es: ",sensibilidad)
especificidad=VN/(VN+FP)
print("La especificidad del modelo es: ",especificidad)
precision=VP/(VP+FP)
print("La precision del modelo es: ",precision)
TasaError=(FP+FN)/(VP+VN+FP+FN)
print("La tasa de error del modelo es: ",TasaError)
PredNeg=VN+FP
print("La prediccion negativa del modelo es: ",PredNeg)
PredPos=VP+FN
print("La prediccion positiva del modelo es: ",PredPos)






Empty DataFrame
Columns: [salary, age, height, weight, games_played_0y1, games_played, games_started, minutes_per_game, assists, blocks, steals, 369.5662921348315]
Index: []

La matriz de confusion es:
 [[247   8]
 [ 12 178]]
La exactitud del modelo es:  0.9550561797752809
La sensibilidad del modelo es:  0.9368421052631579
La especificidad del modelo es:  0.9686274509803922
La precision del modelo es:  0.956989247311828
La tasa de error del modelo es:  0.0449438202247191
La prediccion negativa del modelo es:  255
La prediccion positiva del modelo es:  190


#**CLUSTERING**

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

#Se procede con la implementacion de modelo
from sklearn.cluster import KMeans
from sklearn.metrics import confusion_matrix

In [ ]:
#Este paso es cargar la ruta de mi documento SUPER IMPORTANTE
nxl="/content/drive/MyDrive/SEMESTRE-6/4. BD2_NBA_Data_M_Int.xlsx"

#Antes de usar el dropna se seleccionan las variables para no perder info
XDB = pd.read_excel(nxl, sheet_name=0)  # Cargar la base de datos
XDB = XDB[["salary", "age", "height", "weight", "games_started", "minutes_per_game", "assists", "blocks", "steals", "games_played_0y1"]]
XDB = XDB.dropna()

#Para mirar correlaciones y las variables importantes
plt.figure() #Para enmarcarlo en un cuadro, o pues que se vea lindo
sns.pairplot(XDB,hue="games_played_0y1", size=4, vars=["salary", "age", "height", "weight", "games_started", "minutes_per_game", "assists", "blocks", "steals", "games_played_0y1"]) #Aqui quiero que me explique la variable PreApr frente a las demás
plt.show()

#Aquí en la grafica se puede ver como la variable ingresos y la variable de salida estan correlacionadas, la variable que mejor explica es la que estan mas separadas

#¿Y cuales son las correlaciones? IMPORTANTE ESTE MAPA
plt.figure()
sns.heatmap(XDB.corr(), annot=True)
plt.show()

# Seleccionar las variables de entrada
XD = XDB[["salary", "age", "height", "weight", "games_started", "minutes_per_game", "assists", "blocks", "steals"]]

# Crear cinco clusters
np.random.seed(42)
nc = 5
mkm = KMeans(n_clusters=nc, random_state=42)
mkm.fit(XD)
centroides = mkm.cluster_centers_
print("Los centroides son:\n", centroides)

# Etiquetas de los clusters
labels = mkm.labels_
XDB["Labels"] = labels
print("Los labels son:\n", labels)
display(XDB)

# Caracterización de los clusters
ND = np.bincount(labels)
print("El tamaño de los clusters es:\n", ND)

# Agrupar por cluster y calcular estadísticas
NAboveAvg = XDB.groupby("Labels").agg({"games_played_0y1": lambda x: (x > XDB["games_played_0y1"].mean()).sum()})
NBelowAvg = XDB.groupby("Labels").agg({"games_played_0y1": lambda x: (x <= XDB["games_played_0y1"].mean()).sum()})
print("El número de jugadores por encima del promedio por cluster es:\n", NAboveAvg)
print("El número de jugadores por debajo del promedio por cluster es:\n", NBelowAvg)

# Calcular porcentajes
NAboveAvg["games_played_0y1"] = NAboveAvg["games_played_0y1"] / ND
NBelowAvg["games_played_0y1"] = NBelowAvg["games_played_0y1"] / ND
print("El porcentaje de jugadores por encima del promedio por cluster es:\n", NAboveAvg)
print("El porcentaje de jugadores por debajo del promedio por cluster es:\n", NBelowAvg)






In [ ]:
# Pronóstico
ydp = np.zeros((len(XD), 1))
Vapr = np.array(NAboveAvg)
Vneg = np.array(NBelowAvg)

for k in range(len(XD)):
    napr = labels[k]  # Número de aprobación
    print("El jugador:", k)
    print("El cluster al que pertenece es:", napr)
    print("El porcentaje de Aprobación es:\n", Vapr[napr])
    print("El porcentaje de Rechazo es:", Vneg[napr])

    if Vapr[napr] > Vneg[napr]:
        ydp[k] = 1
    else:
        ydp[k] = 0

# Evaluar el modelo
#Se evalua el modelo
cm=confusion_matrix(yd, ydp)
print("La matriz de confusión es:\n", cm)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]
Exactitud=(VP+VN)/(VP+VN+FP+FN)
Sensibilidad=VP/(VP+FN)
Especificidad=VN/(VN+FP)
Precision=VP/(VP+FP)
Tasaerror=(FP+FN)/(VP+VN+FP+FN)
PredNeg=VN+FP
print("La exactitud es:", Exactitud)
print("La sensibilidad es:", Sensibilidad)
print("La especificidad es:", Especificidad)
print("La precisión es:", Precision)
print("La tasa de error es:", Tasaerror)
print("La tasa de predicción negativa es:", PredNeg)

# Crear modelo de clustering con 5 clusters
np.random.seed(42)
kmeans = KMeans(n_clusters=5, random_state=42)
kmeans.fit(X)

# Evaluar la nueva persona
new_person = np.array([[5000000, 37, 82, 245, 45, 17.8, 0.5, 1.4, 0.4]]).reshape(1, -1)
cluster_prediction = kmeans.predict(new_person)[0]

# Mostrar resultado
print("El jugador pertenece al cluster:", cluster_prediction)
df["Cluster"] = kmeans.labels_
cluster_salary_avg = df.groupby("Cluster")["salary"].mean()
highest_salary_cluster = cluster_salary_avg.idxmax()

# Mostrar resultados
print("El jugador pertenece al cluster:", cluster_prediction)
print("El cluster con el salario más alto es:", highest_salary_cluster)


El jugador: 0
El cluster al que pertenece es: 3
El porcentaje de Aprobación es:
 [0.21495327]
El porcentaje de Rechazo es: [0.78504673]
El jugador: 1
El cluster al que pertenece es: 4
El porcentaje de Aprobación es:
 [1.]
El porcentaje de Rechazo es: [0.]
El jugador: 2
El cluster al que pertenece es: 1
El porcentaje de Aprobación es:
 [0.47787611]
El porcentaje de Rechazo es: [0.52212389]
El jugador: 3
El cluster al que pertenece es: 1
El porcentaje de Aprobación es:
 [0.47787611]
El porcentaje de Rechazo es: [0.52212389]
El jugador: 4
El cluster al que pertenece es: 1
El porcentaje de Aprobación es:
 [0.47787611]
El porcentaje de Rechazo es: [0.52212389]
El jugador: 5
El cluster al que pertenece es: 2
El porcentaje de Aprobación es:
 [0.75438596]
El porcentaje de Rechazo es: [0.24561404]
El jugador: 6
El cluster al que pertenece es: 1
El porcentaje de Aprobación es:
 [0.47787611]
El porcentaje de Rechazo es: [0.52212389]
El jugador: 7
El cluster al que pertenece es: 3
El porcentaje de

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but KMeans was fitted with feature names
  warnings.warn(


#**ANALISIS DE RESULTADOS**
Para este analisis utilizamos tres modelos que me permitieran determinar si el número de juegos jugados por jugador de la NBA está por encima del promedio. Se utilizaron datos de la NBA con variables como salary (salario del jugador), age (edad del jugador), height (altura), weight (peso), games_started (juegos iniciados), minutes_per_game (promedio de minutos jugados por partido), assists (promedio de asistencias por partido), blocks (promedio de bloqueos por partido) y steals (promedio de robos de balón por partido). Además, se agregó una nueva columna games_played_0y1, donde los jugadores por encima del promedio de partidos jugados fueron categorizados con 1 y los que están por debajo con 0.

e implementaron tres modelos: Naive Bayes, Árbol de Decisión y Clustering (K-Means).

El modelo Naive Bayes mostró un desempeño aceptable, con una exactitud del 70.3%. Si bien su precisión alcanzó el 79%, su sensibilidad fue del 41.6%, lo que indica que el modelo tiene dificultades para identificar correctamente a los jugadores con mayor participación en los partidos. Sin embargo, su especificidad del 91.8% refleja que el modelo clasifica con mayor certeza a los jugadores con menor participación.

El Árbol de Decisión fue el modelo más preciso, logrando una exactitud del 95.5%. Este resultado indica que el modelo es altamente eficiente para clasificar correctamente a los jugadores en función de las variables de entrada. Su sensibilidad del 93.7% y especificidad del 96.8% confirman su balance en la identificación de ambas categorías. Además, la precisión del modelo fue del 95.7%, reflejando un excelente desempeño en la predicción de los jugadores con alta participación en los partidos.

Por último, el modelo K-Means permitió segmentar a los jugadores en 5 clusters con base en su rendimiento y características socioeconómicas. La exactitud del modelo fue del 71.2%, con una sensibilidad del 47.3% y una especificidad del 89%. Esto sugiere que el modelo es más efectivo al identificar jugadores con menor tiempo de juego, pero tiene dificultades para clasificar correctamente a los jugadores con mayor participación.

Modelo de Clusterización

Se aplicó el algoritmo K-Means con 5 clusters:

Cluster 0: Jugadores con bajo tiempo de juego y bajo salario.

Cluster 1: Jugadores de experiencia media con contribución equilibrada.

Cluster 2: Estrellas con alto tiempo de juego y salarios elevados.

Cluster 3: Jugadores con poca participación pero alta experiencia.

Cluster 4: Rookies o jugadores con menor impacto estadístico.

Los centroides mostraron que minutes_per_game es clave en la clasificación de los jugadores.

Conclusiones

El análisis mostró que minutes_per_game es la variable más relevante para clasificar a los jugadores. Los jugadores con mayores salarios y experiencia tienden a pertenecer a los clusters de mejor desempeño. El Árbol de Decisión demostró ser el modelo más preciso con un 95% de exactitud, mientras que el modelo de Clustering y Naive Bayes presentaron desempeños aceptables pero con limitaciones en la clasificación de ciertos grupos de jugadores. La clusterización permitió identificar diferentes perfiles de jugadores y entender mejor su impacto en la liga, proporcionandonos asi información valiosa para futuras decisiones estratégicas en equipos y análisis de sus rendimiento.


In [ ]:
import numpy as np
import pandas as pd

# Cargar datos
nxl = "/content/drive/MyDrive/SEMESTRE-6/4. BD2_NBA_Data_M_Int.xlsx"
df = pd.read_excel(nxl)
probabilidad_mas_alta = df["games_played_0y1"].mean() + np.std(df["games_played_0y1"])
intervalo_probabilidad = (df["games_played_0y1"].mean(), probabilidad_mas_alta)

print("Intervalo:", intervalo_probabilidad)


Intervalo: (0.42696629213483145, 0.9216036113046067)
